addons/account/wizard/account_invoice_send.py

```diff
@@ -89,12 +89,20 @@ def _compute_invoice_without_email(self):
 
     def _send_email(self):
         if self.is_email:
-            # with_context : we don't want to reimport the file we just exported.
-            self.composer_id.with_context(no_new_invoice=True, mail_notify_author=self.env.user.partner_id in self.composer_id.partner_ids).send_mail()
             if self.env.context.get('mark_invoice_as_sent'):
                 #Salesman send posted invoice, without the right to write
                 #but they should have the right to change this flag
                 self.mapped('invoice_ids').sudo().write({'is_move_sent': True})
+                # Current transaction might take long time, because in case of
+                # batch invoices we send the emails immediately. But blocking
+                # `is_move_sent` field for a long time might easily lead to
+                # SerializationFailure error, because someone else might change
+                # that field while emails are being sent. SerializationFailure leads
+                # to repeating the process and hence to sending the emails
+                # again.
+                self.env.cr.commit()
```

Comment:
@rco-odoo this is the SerializationFailure error I was talking about recently

Reply:
This is really a corner case, I don't know how it is handled in `mail` and maybe we should investigate that. But this seems overkill.

Reply:
Maybe always use `composition_mode` equal to `comment`? I don't understand why for many records we send mails immediately, while for a single mail it's queued

https://github.com/odoo/odoo/blob/0aff8bb9484a23c0d875d3b12259dd4e0d51e716/addons/account/wizard/account_invoice_send.py#L36

Reply:
What about testing?  I suggest to avoid committing the transaction in a test environment.
All you need is an `if` like https://github.com/odoo/odoo/blob/1f81d60de66cdbe5bfc8c3ef35ddf23b0772d316/addons/crm/models/crm_lead.py#L1813.

Reply:
> This is really a corner case, I don't know how it is handled in `mail` and maybe we should investigate that. But this seems overkill.

@odoo/rd-sm ?

Reply:
Don't know what you really expect from me ... seems this is more linked to accounting than to mail (where we just send emails, with an optional auto_commit parameter). We don't support partial queuing in mail currently, this is done by mass_mailing actually. Having something that depends on mass mailing job of composer is effectively risky, as you might have to wait quite long (hence better to use crons for that kind of job). I suppose your fix is correct then.

Reply:
> What about testing? I suggest to avoid committing the transaction in a test environment. All you need is an `if` like
> 
> https://github.com/odoo/odoo/blob/1f81d60de66cdbe5bfc8c3ef35ddf23b0772d316/addons/crm/models/crm_lead.py#L1813
> .

This is triggering a security warning... Shouldn't there be a better way?
Also, there is a lot of different standards for this (for the upgrade CI, for EDIs, for some other modules,...)

Reply:
Security warning does not mean we cannot do it ^^ I already used that kind of code in several addons.

Reply:
I know that it doesn't mean we can't use it. But we shouldn't need a security review for something like this.
Maybe there should be helpers; maybe commit should never commit in tests unless explicitly specified; maybe there should be a completely safe way of checking if we are in a test;...